### ISSUE TITLE
Question: why System.Void isn't accessible from C#?

### ISSUE THREAD

**Issue by Opiumtm** [dotnet/roslyn#14424]
This code
```C#
var l = new List<System.Void>();
```
don't compile with error
>System.Void cannot be used from C# -- use typeof(void) to get the void type object

Well, `List<System.Void>` is meaningless. But there are scenarios (functional, for examplle) when value type meaning literally "nothing" is needed.

Imagine this:
```C#
public interface IActionWithResult<T>
{
    Task<T> Execute();
}
```
and class, implementing this interface
```C#
// we don't return any result as there is no result
public class MyActionWithoutResult : IActionWithResult<System.Void>
{
    public Task<System.Void> Execute();
}
```
this wouldn't compile with the same error
>System.Void cannot be used from C# -- use typeof(void) to get the void type object

Because of it there are different delegate types and many other - `Action<T>` and `Func<T1, T2>`
If C# have similar to [F# Unit](https://docs.microsoft.com/en-us/dotnet/articles/fsharp/language-reference/unit-type) built-in type, there will be no need for separate `Action` and `Func` types.

On CLR level `void` type seems to be of value type `System.Void`, but this type (meaning literally "nothing" value) isn't accessible from C#. Why?

Is it CLR limitation?

**Comment by svick** [comment]
> Is it CLR limitation?

Yes, [ECMA 335 §II.9.4 Instantiating generic types](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf#page=157&zoom=auto,87,626) says:

> The following kinds of type cannot be used as arguments in instantiations (of generic types or 
methods): 

> * …
> * void (e.g., `List<System.Void>` is invalid)

---

Also, https://github.com/dotnet/roslyn/issues/234 is closely related to this.

**Comment by axel-habermaier** [comment]
It's a CLR limitation. For instance, the following throws an ``ArgumentException`` explaining that the "type 'System.Void' may not be used as a type argument."

```csharp
var t = typeof(List<>);
var vt = t.MakeGenericType(typeof(void));
```

Indeed, that's why F# introduced the ``Unit`` type to avoid issues such as ``Func`` and ``Action`` or ``Task`` and ``Task<T>``.

**Comment by ufcpp** [comment]
related to 0-tuples in https://github.com/dotnet/roslyn/issues/10429

**Comment by dsaf** [comment]
@axel-habermaier indeed:

https://github.com/dotnet/roslyn/issues/10429
![image](https://cloud.githubusercontent.com/assets/3845650/19302644/e198c216-905c-11e6-9c88-5407670b7016.png)

**Comment by Opiumtm** [comment]
@dsaf Empty struct `ValueTuple` isn't intuitive. It is better to name "nuple" as `Unit` or `Nothing` indeed.
